src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageReadmeControl.xaml.cs

```diff
@@ -22,16 +25,47 @@ public partial class PackageReadmeControl : UserControl
 #pragma warning disable CS0618 // Type or member is obsolete
         private IMarkdownPreview _markdownPreview;
 #pragma warning restore CS0618 // Type or member is obsolete
-
+        private bool _isInitializing = false;
         public PackageReadmeControl()
         {
             InitializeComponent();
-            _markdownPreview = MarkdownPreviewSingleton.GetInstance();
-            descriptionMarkdownPreview.Content = _markdownPreview.VisualElement;
         }
 
         public ReadmePreviewViewModel ReadmeViewModel { get => (ReadmePreviewViewModel)DataContext; }
 
+        private async Task InitializeAsync()
+        {
+            if (_markdownPreview is null && !_isInitializing)
+            {
+                _isInitializing = true;
+                await TaskScheduler.Default;
+                IEditorOptionsFactoryService editorOptionsFactoryService;
+                try
+                {
+                    var componentModel = await AsyncServiceProvider.GlobalProvider.GetComponentModelAsync();
```

Comment:
Please plumb through a cancellation token - this will block VS on possible very long running call. I don't know where GetComponentModelAsync comes from, but I added a `CancellationToken` overload to `GetServiceAsync` in 17.11.

Reply:
Note, when you do this, you are inevitably going to run into this [method](https://github.com/NuGet/NuGet.Client/blob/4e10b1165316e09b6570e51b4316d88b493889e9/src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceProviderExtensions.cs#L38).

What is states is no longer true in 17.11+ and it can be deleted. That method is now free-threaded if the service does not need the UI thread. You should only be calling [ServiceExtensions.GetServiceAsync](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.serviceextensions.getserviceasync?view=visualstudiosdk-2022) overloads, the docs are up-to-date.

Reply:
I switched this to use https://github.com/NuGet/NuGet.Client/blob/c7a7a1ec509590a4eb2530d72fe266a2feb5008a/src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs?plain=1#L85C1-L85C108.

Reply:
I'll clean-up the extra method. https://github.com/NuGet/Home/issues/14007